Volzhanskiy

NikolayVolzhanskiy opened the pull request.

NikolayVolzhanskiy committed:
Update LightningTree

Заполнил функции вычисления тока, потенциала и сигмы

NikolayVolzhanskiy committed:
update LightningTree

Переписал с учётом того, что обсудили

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -1,34 +1,37 @@
-    double Q_plus_s, double Q_minus_s, double resistance,
-    double E_plus, double E_minus, double alpha, double beta,
-    double sigma,
+    double _h, double _delta_t, double _r, double _R,
```
Не надо менять названия переменных. В конструкторе класса разрешается для аргументов использовать имена полей

NikolayVolzhanskiy replied:
Хорошо

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -39,35 +42,179 @@ void LightningTree::NextIter() {
+            double mirror_l = abs(sqrt(pow((vertex.coords[0] - point.coords[0]), 2) + pow((vertex.coords[1] - point.coords[1]), 2) + pow((vertex.coords[2] + point.coords[2]), 2)));
+            if (l < kEps)
+            {
+                Phi = vertex.q / (4 * M_PI * epsilon_0) * (1 / (h + r) + 1 / (h + mirror_l + r)) + 
```
Потенциал точки должен суммироваться в цикле. В твоем случае потенциал будет равен вкладу последнего рассмотренного заряда

NikolayVolzhanskiy replied:
Поправлю

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -39,35 +42,179 @@ void LightningTree::NextIter() {
+                Phi = vertex.q / (4 * M_PI * epsilon_0) * (1 / (h + r) + 1 / (h + mirror_l + r)) + 
+                            vertex.Q / (4 * M_PI * epsilon_0) * (1 / (h + R) + 1 / (h + mirror_l + R));
+            }
+            Phi = vertex.q / (4 * M_PI * epsilon_0) * (1 / (l + r) + 1 / (mirror_l + r)) +
```
зеркальные заряды берутся с противоположным знаком и нет смысла рассматривать для них отдельно случай когда l < kEps, так как mirrol_l > kEps всегда

NikolayVolzhanskiy replied:
Так такой проверки и нет. Проверка только на узлы, а зеркало считается просто за счёт смены знака координаты по оси z

Step2Victory replied:
Под ифом вклад зеркального заряда должен считаться также как и вне ифа

Step2Victory commented on src/LightningTree.h:
```diff
@@ -1,6 +1,14 @@
+#define epsilon_0 8.854187817619999806e-12
+#define kEps 1e-9
+#define _USE_MATH_DEFINES
```
луяше использовать constexpr вместо define. Не очень понял зачем здесь директива _USE_MATH_DEFINES

NikolayVolzhanskiy replied:
_USE_MATH_DEFINES для использования констант в библиотеке cmath, хотя можно просто pi прописать.
+ Почему define лучше не использовать?

Step2Victory replied:
В современных стандартах стараются использовать меньше макросов. Им на замену пришли constexpr выражения. Лучше константы использовать так: https://en.cppreference.com/w/cpp/numeric/constants

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -39,35 +42,179 @@ void LightningTree::NextIter() {
+    Описание метода. Узнать что делает этот метод!
+    */
     // TO DO
+    std::unordered_map<VertexPtr, std::pair<double, double>> delta_charges;
```
В новой реализации нет смысла в мапе

NikolayVolzhanskiy replied:
Да я не знаю пока чем заменить это. Может просто вектор из массива с 3я элементами использовать...

Step2Victory replied:
Не надо пытаться заменить. Я использовал эту мапу в своей реализации, так как прям по проходу в цикле считал потенциалы. Тоесть если бы я менял заряд прямо в цикле, то на следующей итерации я бы посчитал потенциал с учетом этого заряда, а надо с учетом старого значения. Сейчас можно прямо в цикле менять заряды, так как потенциалы уже посчитаны заранее

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -39,35 +42,179 @@ void LightningTree::NextIter() {
+    */
     // TO DO
+    std::unordered_map<VertexPtr, std::pair<double, double>> delta_charges;
+    for (auto& elem : graph)
```
здесь надо простой пройтись по всем вершинам в графе и перебрать соседей. Посчитать сколько заряда перейдет текущей вершины от каждого соседа (сумма таких зарядов и будет delta q для текущего заряда)

NikolayVolzhanskiy replied:
это я переписал, только не скинул ещё

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -39,35 +42,179 @@ void LightningTree::NextIter() {
+        delta_charges[vertex].first -= CurrentSheath(vertex);
+        delta_charges[vertex].second += CurrentSheath(vertex);
+    }
+    for (auto& elem : delta_charges)
```
это тоже не нужно

NikolayVolzhanskiy replied:
дополнительный цикл?
Не совсем понял

Step2Victory replied:
delta_charges не нужны. Делай += в предыдущем цикле

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -39,35 +42,179 @@ void LightningTree::NextIter() {
+                    std::array<double, 3> coords = {0, 0, 0}
+                    countCoords(coords, vertex_id, std::vector{ i, j, k });
+                    LightningTree::Vertex new_vertex = { 0, 0, 0, coords, 0 };
+                    vertices.push_back(new_vertex);
```
эту строку перенести под if, а pop_back  убрать

NikolayVolzhanskiy replied:
Не. так не получится. У нас же индекс вершины везде фигурирует и если мы вершину не добавим, то её сложно использовать будет. А после всех проверок, если ничего не создалось, просто удаляем вершину, т.к. к ней не ведут ребра

Step2Victory replied:
Окей, согласен. Тогда лучше GrowthCriterion(const Vertex&, const Vertex&). Так будет логичнее, так как ребро следует создать, когда критерий выполнен. Если критерий не выполнен, то манипуляции с полями графа производить не надо

Step2Victory reviewed (changes requested):
Набросал комментариев по коду

NikolayVolzhanskiy committed:
Update LightningTree.cpp

NikolayVolzhanskiy committed:
update LightningTree

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -54,25 +54,27 @@ void LightningTree::CountPotential() {
         }
-        point.Phi = Phi + phi_a->getValue(point.coords);
+        // ???
+        point.Phi = Phi + phi_a->getValue(point.coords); 
```
У phi_a нет метода get_value. Вызывай phi_a(point.coords)

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -177,16 +181,19 @@ void LightningTree::Grow() {
+
+                    std::array<double, 3> coords = {0, 0, 0};
                     countCoords(coords, vertex_id, std::vector{ i, j, k });
                     LightningTree::Vertex new_vertex = { 0, 0, 0, coords, 0 };
```
Лучше писать new_vertex = {.q=0, .Q = 0, ,phi = 0, .coords=coords, .grow_iter_number=0}

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -1,29 +1,29 @@
 #include "LightningTree.h"
 
```
namespace LightningTree {
...
};

NikolayVolzhanskiy replied:
Ругается на все методы внутри пространства

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -177,16 +181,19 @@ void LightningTree::Grow() {
+                        LightningTree::Edge new_edge = { vertex_id, vertices.size() - 1, 0, 0};
                         edges.push_back(new_edge);
                         graph[v][i][j][k] = edges.size() - 1;
-                        temp_graph.push_back(CreateNode(vertices.size() - 1), edges.size() - 1, std::vector{i, j, k});
```
пометить, что vertex_id не перефирийная и grow_less = 0

Step2Victory commented on src/LightningTree.h:
```diff
@@ -60,15 +61,15 @@ class LightningTree {
     double alpha;
     double beta;
     double sigma;
-    size_t periphery_size;
```
добавить два вектора bool (один для перифирийных, другой для удаленных)

Step2Victory commented on src/LightningTree.h:
```diff
@@ -1,15 +1,16 @@
 
+constexpr double epsilon_0 = 8.854187817619999806e-12;
+constexpr double kEps = 1e-9;
+constexpr double PI = 3.14159265358979323846;
```
#include <numbers>

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -206,9 +213,10 @@ void LightningTree::Delete() {
 
```
std::vector<bool> is_perephery;
for (id : vertices) {
    if (DeletionCriteria(id)) {
удалить вершину (пометить удаленной)
}

}

NikolayVolzhanskiy committed:
update LightningTree

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -54,18 +56,18 @@ void LightningTree::CountPotential() {
+                Phi += vertex.q / (4 * numbers::pi * epsilon_0) * (1 / (h + r) + 1 / (mirror_l + r)) + 
+                            vertex.Q / (4 * numbers::pi * epsilon_0) * (1 / (h + R) + 1 / (mirror_l + R));
             }
             Phi += vertex.q / (4 * PI * epsilon_0) * (1 / (l + r) + 1 / (mirror_l + r)) +
```
здесь тоже заменить pi

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -205,15 +216,37 @@ void LightningTree::Grow() {
+        if(DeletionCriterion(v_id))
+        {
+            vertices_activity[v_id] = false;
+            size_t graph_id = finde_index(v_id);
```
что делает finde_index?

NikolayVolzhanskiy replied:
ищет узел в графе, связанный с индексом вершины, и возвращает его index в векторе.
Сам метод не написал ещё

Step2Victory commented on src/LightningTree.cpp:
```diff
@@ -205,15 +216,37 @@ void LightningTree::Grow() {
+                    {
+                        if (graph[graph_id][i][j][k] == -1 && (i == 1 && j == 1 && k == 1)) continue;
+                        edges_activity[graph[v][i][j][k]] = false;
+                        if(vertices[edges[graph[v][i][j][k]].from].growless_iter_number /*Проверка на количество исходящих ребер*/)
```
growless_iter_number используется только в критерии удаления

Step2Victory replied:
здесь надо проверить что степень вершины == 1

NikolayVolzhanskiy replied:
я подумал использовать этот параметр для 2х критериев. Как количество исходящих ребер у вершины и количество раз, когда из вершины не выросло ребро, только с отрицательным знаком.

Step2Victory commented on src/LightningTree.h:
```diff
@@ -61,18 +61,21 @@ class LightningTree {
+    std::vector<bool> vertices_peripherality;
+    std::vector<bool> vertices_activity;
     std::vector<Edge> edges;
+    std::vector<bool> edges_activity;
```
кстати реинкарнация ребер интересный момент. Можно потом спросить у физиков

NikolayVolzhanskiy replied:
так проще будет реализовать. Если вершины могут вернуться, то и ребра то же, хоть ребра и могут быть из другой вершины, возможность всё равно есть

NikolayVolzhanskiy committed:
Fix changes Lightningtree

NikolayVolzhanskiy committed:
Update Lightningtree

Использовал переменную grow_iter_number как количество выросших рёбер и количество пропусков роста.
Добавил в delete вырезание узла графа.

NikolayVolzhanskiy committed:
Merge branch 'main' into Volzhanskiy
